### PR TITLE
User Registration Web Handler

### LINKED ISSUE #152: User registration API
## Task Summary
- User registration API using which user can be registered.
- API will accept all necessary json data for registration through HTTP request and return appropriate json response.

## Acceptance Criteria
- [x] API must accept data for registration
- [x] Registration with given data only.
- [x] Duplicate registration should not happen
- [x] Handle username (Auto generate/ Accept username), but it must be unique
- [x] Send response with success or failure status. 

## Subtasks
- NA.
  
## Related Issues
- NA

## Additional Notes
- Basic and necessary validation must be handled, _such as unique `username`_

### PR DESCRIPTION
**Pull request by priyapal2000**
## Description

- **Implemented user registration web handler responsible for processing user registration requests.**
- Handles user data submission.
- Creates new user accounts in the database.
- Generates unique user IDs if necessary.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Project Configuration
- [ ] Repository Configuration
- [ ] Documentation update - Code guide/ docs (like, javadoc or API)
- [ ] Other (please describe):

## Screenshots / GIFs / Media (if applicable)

- N/A

## How Has This Been Tested? / Test Plan

- Sent a `POST` request to `/api/auth/register` with user data using the Postman tool.
- Verified responses for success and failure scenarios.

## Checklist

- [x] My code follows the project’s [coding guidelines](./../../docs/guidelines/coding-guides/).
- [x] I have performed a self-review of my own code.
- [x] I have commented on my code where necessary.
- [ ] I have made corresponding changes to the documentation (if applicable).
- [ ] I have added tests that verify the functionality of the code (if applicable).
- [ ] All new and existing tests pass.

## How to Use / Enable

1. Create a registration form to collect user details or manually collect user information.
2. Send the data using an HTTP `POST` request to `/api/user/register`.
   - Endpoint: `/api/user/register`
   - HTTP Method: `POST`
   - Request Body:
     `content-type`:`application/json`
     ```json
     {
       "address": "Jasiwal Nagar",
       "city": "Mumbai",
       "dob": "2000-02-06",
       "email": "priya@gmail.com",
       "moNo": 352928,
       "name": "Priya",
       "pincode": 400097,
       "enabled": false,
       "username": "palpriya"
     }
     ```
   - Response Body:
     `content_type`:`application/json`
     - User added successfully
       Success: `Status: 201 Created`  
       ```json
        {
          "id": 4,
          "username": "palpriya",
          "password": "S6eBeq4S",
          "user": {
            "id": 4,
            "name": "Priya",
            "dob": "2000-02-06",
            "address": "Jasiwal nagar",
            "moNo": "352928",
            "city": "Mumbai",
            "pincode": 400097,
            "username": "palpriya",
            "createdAt": "2024-12-28",
            "enabled": true
          }
        }
       ```
     - User already exists
       Failure: `Status: 422 Unprocessable Entity`  
       ```json
        {
          "username": "pihu321",
          "error": "User already exists."
        }
       ```
     - Invalid input (e.g., username is null)  
       Failure: `Status: 400 Bad Request`  
       ```json
        {
          "error": "Missing user details!, Check {Username}, {Name}, {City}, {Mobile No}, {Pincode}."
        }
       ```
     - Unable to create Auth entry
       Failure: `Status: 503 Service Unavailable.
       ```json 
        {
          "username": "pihu321",
          "error": "Failed to create new authorized user."
        }
       ```

## Additional Notes
- N/A

## Closing Issues
Closes #152

### CODE REVIEW

**Comment by priyapal2000** [comment]
@ajayg2808 
Please review the PR.

**Comment by priyapal2000** [comment]
@ajayg2808 
Please review the PR.

**Comment by ajayg2808** [comment]
@priyapal2000,
Do below updates:
- [x] Correct section: **How Has This Been Tested?/ Test Plan**
- [x] Mention endpoints: **How to use/ enable**
- [x] Use JSON code block for Request and Response body currently it is normal code block
- [x] Handle failure properly do not delegate it to `INTERNAL_SERVER_ERROR`
- [x] In the Request Body `enabled` is of type `boolean` and not a `int`
- [x] Assign PR

**Comment by priyapal2000** [comment]
@ajayg2808 
PR is updated

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import com.arcitech.model.User;
+import com.arcitech.model.UserAuth;
+import com.arcitech.service.UserService;
+
+/**
+ * @author Priya
+ * 
+ */
+@Controller
+public class UserRegistrationController {
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Class name should be `UserController`.

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	
+	@Autowired
+	private UserService userService;
+	
+	@GetMapping("findByUserId/{id}")
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Do not implement `Get/ Find User` API

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		if(user==null) {
+			return new ResponseEntity<>(user, HttpStatus.OK);
+		}
+		else {
+			return new ResponseEntity<>(user, HttpStatus.NOT_FOUND);
+			}
+	}
+	// Create a new user
+	@PostMapping("/adduser")
+	public ResponseEntity<UserAuth> createUser(@RequestBody User user) {
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Change endpoint

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		// Call the service method to add the user
+		UserAuth userAuth = userService.addUser(user);
+		// If user creation was successful, return 201 Created, else return 500 Internal Server Error
+	    if (userAuth!=null) {
+	        return new ResponseEntity<>(userAuth, HttpStatus.CREATED);
+	    }
+	    else {
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Do not create `INTERNAL_SERVER_ERROR`. Internal server error depicts system/ server failure to handle use case.

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	        return new ResponseEntity<>(userAuth, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Return proper JSON response including proper error message "Why user not added?"

File: Arc-i-Tech/src/main/java/com/arcitech/model/PersonalDetails.java [hunk 6]
```diff
@@ -24,7 +24,7 @@ public abstract class PersonalDetails extends CommonFields {
 	@Column(unique = true, nullable = false)
 	private int moNo;
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/model/PersonalDetails.java]: Change `moNo` field to `String`

File: Arc-i-Tech/src/main/java/com/arcitech/repository/UserAuthRepository.java
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/repository/UserAuthRepository.java:1]: Use/ Copy it from #191

File: Arc-i-Tech/src/main/java/com/arcitech/repository/UserRepository.java
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/repository/UserRepository.java:1]: Use/ Copy it from #191

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * 
+ */
+public interface UserService {
+	public UserAuth addUser (User user);
+
+	/**
+	 * @param id
+	 * @return
+	 */
+	public User getUserById(Long id);
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/service/UserService.java]: Why it is needed?

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+import com.arcitech.model.UserAuth;
+import com.arcitech.repository.UserAuthRepository;
+import com.arcitech.repository.UserRepository;
+
+/**
+ * @author Priya
+ * 
+ */
+@Service
+public class UserServiceImplement implements UserService {
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: Rename class to `UserServiceImpl` and move it to package `com.arcitech.service.impl`

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+	@Autowired
+	UserRepository userRepository;
+	
+	@Autowired
+	UserAuthRepository userAuthRepository;
+	
+	private static final Logger logger = LoggerFactory.logger(UserServiceImplement.class);
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: Do not use logger, we are not started using it yet.

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	@Autowired
+	UserRepository userRepository;
+	
+	@Autowired
+	UserAuthRepository userAuthRepository;
+	
+	private static final Logger logger = LoggerFactory.logger(UserServiceImplement.class); 
+
+	@Override
+	public UserAuth addUser(User user) {
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: Add user must return `User` object only.

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	        return userAuth;
+	    } catch (Exception e) {
+	    	// Log the error and return null in case of failure
+	    	logger.error("Error saving user: " + e.getMessage());
+	    	return null;
+	    }
+	}
+	
+	@Override
+	public User getUserById(Long id) {
```
Comment (ajayg2808) [inline Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: Not needed.

**Review by ajayg2808 — CHANGES_REQUESTED** [review]
- [x] Endpoint `/api/auth/register` is not found
- [x] Endpoint mentioned in pull request description and actual implemented endpoint is different.
- [x] Endpoint should be `/api/user/register` or `/api/user/add`

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java (hunk 1 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Is done.

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java (hunk 2 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: Use http request Post.

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java (hunk 3 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: is done.

File: Arc-i-Tech/src/main/java/com/arcitech/model/PersonalDetails.java (hunk 6 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/model/PersonalDetails.java]: changes is done.

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java (hunk 9 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: is done.

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java (hunk 11 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: ok

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java (hunk 4 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: is done.

Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: is done

File: Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java (hunk 5 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/controllers/UserRegistrationController.java]: is done.

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java (hunk 8 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: Done

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserService.java (hunk 7 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/service/UserService.java]: ok

**Comment by ajayg2808** [comment]
@priyapal2000,
These conversations are not resolved yet:
- [x] https://github.com/Arc-i-Tech/Arc-i-Tech.github.io/pull/209#discussion_r1897231906
- [x] https://github.com/Arc-i-Tech/Arc-i-Tech.github.io/pull/209#discussion_r1897232668
- [x] https://github.com/Arc-i-Tech/Arc-i-Tech.github.io/pull/209#discussion_r1897234032

File: Arc-i-Tech/src/main/java/com/arcitech/repository/UserAuthRepository.java
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/repository/UserAuthRepository.java:1]: it's done

File: Arc-i-Tech/src/main/java/com/arcitech/repository/UserRepository.java
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/repository/UserRepository.java:1]: done.

File: Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java (hunk 10 above)
Comment (priyapal2000) [reply Arc-i-Tech/src/main/java/com/arcitech/service/UserServiceImplement.java]: is done

**Comment by priyapal2000** [comment]
@ajayg2808 
All Conversations is resolve.

**Comment by priyapal2000** [comment]
@ajayg2808 
PR is updated .

**Review by ajayg2808 — APPROVED** [review]

**Pull request merged**